hathor/indexes/utxo_index.py

```diff
@@ -0,0 +1,326 @@
+from hathor.transaction import BaseTransaction
+from hathor.transaction.scripts import parse_address_script
+from hathor.util import peekable, sorted_merger
+
+logger = get_logger()
+settings = HathorSettings()
+
+
+@dataclass(frozen=True)
+class UtxoIndexItem:
+    token_uid: bytes
+    tx_id: bytes
+    index: int
+    address: str
+    amount: int
+    timelock: Optional[int]
+    heightlock: Optional[int]
+
+    def __repr__(self):
+        return (
+            'UtxoIndexItem('
+            f'token_uid={self.token_uid.hex()},'
+            f'tx_id={self.tx_id.hex()},'
+            f'index={self.index},'
+            f'address={self.address},'
+            f'timelock={self.timelock},'
+            f'heightlock={self.heightlock}'
+            ')'
+        )
+
+
+class UtxoIndex(BaseIndex):
+    """ Index of UTXOs by address+token
+
+    This index is currently optional and only used for an optional API.
+
+    It is expected to be called with only non-voided transactions and blocks, including the genesis. It ignores
+    token-creation transactions and data-only outputs.
+    """
+
+    def __init__(self):
+        self.log = logger.new()
+
+    # interface methods provided by the base class
+
+    def init_loop_step(self, tx: BaseTransaction) -> None:
+        tx_meta = tx.get_metadata()
+        # we don't want ANY type of voided transaction
+        if not tx_meta.voided_by:
+            self.add_tx(tx)
+
+    def add_tx(self, tx: BaseTransaction) -> None:
+        """ This method processes both inputs and output from a transaction.
+
+        - if the transaction has already been added, warning is logged and inputs/outputs aren't processed
+        - mark transaction as added
+        - inputs are removed from the index
+        - outputs are added to the index
+        """
+        tx_meta = tx.get_metadata()
+        if tx_meta.voided_by:
+            self.log.debug('skip add utxos of voided tx', tx=tx.hash_hex)
+            return
+        self.log.debug('add utxos of', tx=tx.hash_hex)
+        heightlock: Optional[int]
+        timelock: Optional[int]
+        assert tx.hash is not None
+        # ignore if already in the index
+        if self._is_in_index(tx.hash):
+            self.log.warn('already in UTXO index, skip add', tx=tx.hash_hex)
+            return
+        self._add_tx(tx.hash)
+        tx_meta = tx.get_metadata()
+        # inputs are spent, remove them from the index
+        for tx_input in tx.inputs:
+            spent_tx = tx.get_spent_tx(tx_input)
+            spent_tx_meta = spent_tx.get_metadata()
+            spent_tx_output = spent_tx.outputs[tx_input.index]
+            address_script = parse_address_script(spent_tx_output.script)
+            if not address_script:
+                # XXX: completely ignore scripts that aren't "addressable"
+                continue
+            if spent_tx.is_block:
+                # XXX: timelock forced to None when there is a heightlock
+                timelock = None
```

Comment:
That's an interesting case. Why don't we add the transaction to both lists?

Reply:
We definitely _can_, and maybe we a complete solution would be doing that, but that creates a problem to avoid duplicated UTXOs, it means we have to check the resulting list for every addition to the list (or at least to the additions that can have a duplication).

IMO we don't lose much in practice by having this restriction and we can make a fully complete solution in the future that properly migrates the existing index.

Reply:
Done: #426

Reply:
Great! Thanks!